plugin/plugin.lua

```diff
@@ -0,0 +1,268 @@
+    -- Note that the target API must always digitally verify the JWT access token
+    if config.allow_tokens then
+        local auth_header = ngx.var.http_authorization
+        if auth_header and string.len(auth_header) > 7 and string.lower(string.sub(auth_header, 1, 7)) == 'bearer ' then
+            return
+        end
+    end
+
+    -- For cookie requests, verify the web origin in line with OWASP CSRF best practices
+    if config.trusted_web_origins then
+
+        local web_origin = ngx.req.get_headers()['origin']
+        if not web_origin or not array_has_value(config.trusted_web_origins, web_origin) then
+            ngx.log(ngx.WARN, 'The request was from an untrusted web origin')
+            unauthorized_request_error_response(config)
+        end
+    end
+
+    -- For data changing requests do double submit cookie verification in line with OWASP CSRF best practices
+    if method == 'POST' or method == 'PUT' or method == 'DELETE' or method == 'PATCH' then
+
+        local csrf_cookie_name = 'cookie_' .. config.cookie_name_prefix .. '-csrf'
+        local csrf_cookie = ngx.var[csrf_cookie_name]
+        if not csrf_cookie then
+            ngx.log(ngx.WARN, 'No CSRF cookie was sent with the request')
+            unauthorized_request_error_response(config)
+        end
+
+        local csrf_token = decrypt_cookie(csrf_cookie, config.encryption_key)
+        if not csrf_token then
+            ngx.log(ngx.WARN, 'Error decrypting CSRF cookie')
+            unauthorized_request_error_response(config)
+        end
+        
+        local csrf_header = ngx.req.get_headers()[get_csrf_header_name(config)]
+        if not csrf_header or csrf_header ~= csrf_token  then
+            ngx.log(ngx.WARN, 'Invalid or missing CSRF request header')
+            unauthorized_request_error_response(config)
+        end
+    end
+
+    -- Next verify that the main cookie was received and get the access token
+    local at_cookie_name = 'cookie_' .. config.cookie_name_prefix .. '-at'
+    local at_cookie = ngx.var[at_cookie_name]
+    if not at_cookie then
+        ngx.log(ngx.WARN, 'No access token cookie was sent with the request')
+        unauthorized_request_error_response(config)
+    end
+
+    ngx.log(ngx.WARN, "*** DEBUG")
+    ngx.log(ngx.WARN, at_cookie)
+
+    -- Decrypt the access token cookie, which is encrypted using AES256
+    local access_token = decrypt_cookie(at_cookie, config.encryption_key)
+    if not access_token then
+        ngx.log(ngx.WARN, 'Error decrypting access token cookie')
+        unauthorized_request_error_response(config)
+    end
+
+    -- Set the request header to supply the access token to the next plugin or the target API
+    ngx.req.set_header('authorization', 'Bearer ' .. access_token)
+
+    -- Clear headers of no interest to the target API
+    if config.remove_cookie_headers then
+        ngx.req.clear_header('cookie')
+        if csrf_header_name then
```

Comment:
Something's odd with this if. Where does `csrf_header_name` come from?

Reply:
Fixed